testing/trino-testing-containers/src/main/java/io/trino/testing/containers/TestContainers.java

```diff
@@ -58,7 +62,40 @@ public static Closeable startOrReuse(GenericContainer<?> container)
         if (reuse) {
             return () -> {};
         }
-        return container::stop;
+        if (container.getStartupAttempts() <= 1) {
+            return container::stop;
+        }
+        return () -> {
+            container.stop();
+            // if there were multiple startup attempts, failed containers need to be manually removed
```

Comment:
if the containers actually terminate then why is the cleanup even needed?

But yes if the containers stick around (e.g. retry due to timeout) then I understand.

Can you please explain what I miss?

Reply:
They're stopped, but not removed. The resource leak detector doesn't allow that

Reply:
> Now I started to wonder if stopped containers should not be considered leaks? Ryuk would clean them up anyway

I tend to agree.

cc: @findepi

Reply:
stopped containers are also a leak (they take up disk space, for example. Unless we  want to remain independent from ryuk for cleanup (which i personally want).

it's good that we have https://github.com/trinodb/trino/pull/21280 as a workaround
ngl, this PR wasn't very pretty. probably a cleaner fix could be done directly in TC. cc @wendigo